rust/scx_utils/src/load_balance.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+
+// This software may be used and distributed according to the terms of the
+// GNU General Public License version 2.
+
+//! # SCX Load Balancer
+//!
+//! A crate providing abstractions for implementing load balancing logic
+//! between scheduling domains.
+//!
+//! Initializing LoadBalancer
+//! -------------------------
+//!
+//! LoadBalancer objects are created from LoadAggregator objects. As the name
+//! suggests, LoadAggregator objects aggregate load passed by the caller, and
+//! create a LoadBalancer object that contains load and weights for each
+//! domain; adjusted for any infeasible weights.
+//!
+//! Note that the abstractions here are meant to be generic across schedulers.
+//! LoadAggregator assumes nothing about the scheduling domains being passed to
+//! it. For example, they may span layers defined in a scheduler, domains
+//! specified by the user via cpumask strings, or domains that span L3 caches.
+//! The crate simply aggregates load according to domain with IDs specified by
+//! the caller, computes global load sums, and adjusts any infeasible weights
+//! such that all infeasible scheduling entities still get their full duty
+//! cycle.
+//!
+//! Assume we're on a 16-core (32-CPU) host with two core complexes:
+//!
+//!```
+//!     let mut aggregator = LoadBalancer::aggregator(32, false);
+//!     aggregator.record_dom_load(0, 1, 1.0);
+//!     aggregator.record_dom_load(1, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(31, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(63, 1, 1.0);
+//!     aggregator.record_dom_load(64, 10000, 1.0);
+//!
+//!     let balancer = aggregator.calculate();
+//!
+//!     // Outputs: 66.06451612903226
+//!     info!("{}", balancer.global_load_sum());
+//!```
+//!
+//! For more details on the infeasible weights problem, please see the following
+//! document:
+//!
+//! https://drive.google.com/file/d/1fAoWUlmW-HTp6akuATVpMxpUpvWcGSAv
+
+use anyhow::bail;
+use anyhow::Result;
+use std::collections::BTreeMap;
+
+const MIN_WEIGHT: usize = 1;
+const MAX_WEIGHT: usize = 10000;
+
+#[derive(Debug)]
+pub struct LoadBalancer {
+    dom_load_sums: Vec<f64>,
+    dom_dcycle_sums: Vec<f64>,
+    global_dcycle_sum: f64,
+    global_load_sum: f64,
+    w_prime: f64,
+}
+
+impl LoadBalancer {
+    /// Create a LoadAggregator object that can be given domains and loads by
+    /// the caller, and then used to create a LoadBalancer object.
+    pub fn aggregator(nr_cpus: usize, dcycle_only: bool) -> LoadAggregator {
+        LoadAggregator {
+            doms: BTreeMap::new(),
+            global_loads: BTreeMap::new(),
+            nr_cpus,
+            max_weight: 0,
+            global_dcycle_sum: 0.0f64,
+            global_load_sum: 0.0f64,
+            w_prime: 10000.0f64,
+            dcycle_only,
+        }
+    }
+
+    /// Return the global, host-wide sum of duty cycles.
+    pub fn global_dcycle_sum(&self) -> f64 {
```

Comment:
Dumb question. What's the benefit of these accessors? Given that the embedding struct contains only data, what's the difference from just allowing users access the members?

Reply:
This way the struct can't be mutated. It felt like a bit of a tighter API to do it this way and force it to be read-only.